packages/langium/src/grammar/type-system/type-collector/types.ts

```diff
@@ -376,6 +385,156 @@ function pushReflectionInfo(node: CompositeGeneratorNode, name: string) {
+    let result = true;
+    if (isPrimitiveType(propertyType)) {
+        if (propertyType.primitive === 'string') {
+            return true;
+        } else {
+            return false;
+        }
+    } else if (isStringType(propertyType)) {
+        return true;
+    } else if (!isPropertyUnion(propertyType)) {
+        return false;
+    } else {
+        for (const type of propertyType.types) {
+            if (isValueType(type)) {
+                if (isUnionType(type.value)){
+                    if (!containsOnlyStringTypes(type.value.type)) {
+                        return false;
+                    }
+                } else {
+                    return false;
+                }
+            } else if (isPrimitiveType(type)) {
+                if (type.primitive !== 'string' || !type.regex) {
+                    return false;
+                }
+            } else if (isPropertyUnion(type)) {
+                result = containsOnlyStringTypes(type);
+            } else if (!isStringType(type)) {
+                return false;
+            }
+        }
+    }
+    return result;
+}
+
+function createDataTypeCheckerFunctionReturnString(subTypes: string[], strings: string[], regexes: string[]): string {
+    let returnString = 'return ';
+
+    const allArray = [
+        ...subTypes.map(e => `is${e}(item)`),
+        ...strings.map(e => `item === '${e}'`),
+        ...regexes.map(e => `/${e}/.test(item)`),
+    ];
+
+    returnString += allArray.join(' || ');
+    return returnString;
+}
+
 function escapeReservedWords(name: string, reserved: Set<string>): string {
     return reserved.has(name) ? `^${name}` : name;
 }
+
+function collectStringValuesFromDataType(propertyType: PropertyType): string[] {
+    const values: string[] = [];
+    if (isStringType(propertyType)) {
+        return [propertyType.string];
+    }
+    if (isPropertyUnion(propertyType)) {
+        for (const type of propertyType.types) {
+            if (isStringType(type)) {
+                values.push(type.string);
+            } else if (isPropertyUnion(type)) {
+                values.push(...collectStringValuesFromDataType(type));
+            }
+        }
+    }
+
+    return values;
+}
+
+function collectRegexesFromDataType(propertyType: PropertyType): string[] {
+    const regexes: string[] = [];
+    if (isPrimitiveType(propertyType) && propertyType.primitive === 'string' && propertyType.regex) {
+        regexes.push(propertyType.regex);
+    }
+    if (isPropertyUnion(propertyType)) {
+        for (const type of propertyType.types) {
+            if (isPrimitiveType(type) && type.primitive === 'string' && type.regex) {
+                regexes.push(type.regex);
+            } else if (isPropertyUnion(type)) {
+                regexes.push(...collectRegexesFromDataType(type));
+            }
+        }
+    }
+    return regexes;
+}
+
+function pushDataTypeReflectionInfoForNumber(node: CompositeGeneratorNode, union: UnionType) {
```

Comment:
I believe we can add another layer of abstraction in there, right now there's a lot of repetition in the `pushDataTypeReflectionFor...` functions.